pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return fmt.Errorf("failed to write recovery config %q: %v", recoveryConfigPath, err)
+	}
+
+	recoveryPod, err := s.recoveryPod()
+	if err != nil {
+		return fmt.Errorf("failed to create recovery pod: %v", err)
+	}
+
+	recoveryPodBytes, err := yaml.Marshal(recoveryPod)
+	if err != nil {
+		return fmt.Errorf("failed to marshal recovery pod: %v", err)
+	}
+
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err = ioutil.WriteFile(recoveryPodManifestPath, recoveryPodBytes, 644)
+	if err != nil {
+		return fmt.Errorf("failed to write recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// Create client cert
+	ca := crypto.CA{
+		Config:          rootCaConfig,
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+	}
+
+	// Create client certificates for system:admin
+	clientCert, err := ca.MakeClientCertificateForDuration(&user.DefaultInfo{Name: "system:admin"}, certValidity)
+	if err != nil {
+		return fmt.Errorf("failed to create client certificate: %v", err)
+	}
+
+	clientCertBytes, clientKeyBytes, err := clientCert.GetPEMBytes()
+
+	s.restConfig = &rest.Config{
+		Host: "https://localhost:7443",
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:   servingCaCertPath,
+			CertData: clientCertBytes,
+			KeyData:  clientKeyBytes,
+		},
+	}
+
+	kubeconfig, err := s.KubeConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create kubeconfig: %v", err)
+	}
+
+	kubeconfigBytes, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to marshal kubeconfig: %v", err)
+	}
+
+	kubeconfigPath := path.Join(recoveryDirPath, AdminKubeconfigFileName)
+	err = ioutil.WriteFile(kubeconfigPath, kubeconfigBytes, 600)
+	if err != nil {
+		return fmt.Errorf("failed to write kubeconfig %q: %v", kubeconfigPath, err)
+	}
+
+	return nil
+}
+
+func (s *Apiserver) WaitForHealthz(ctx context.Context) error {
+	// We don't have to worry about connecting to an old, terminating apiserver
+	// as our client certs are unique to this instance
+
+	kubeClientset, err := s.GetKubeClientset()
+	if err != nil {
+		return err
+	}
+
+	return wait.PollUntil(500*time.Millisecond, func() (bool, error) {
+		req := kubeClientset.RESTClient().Get().AbsPath("/healthz")
+
+		klog.V(1).Infof("Waiting for recovery apiserver to come up at %q", req.URL())
+
+		_, err = req.DoRaw()
+		if err != nil {
+			klog.V(5).Infof("apiserver returned error: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	}, ctx.Done())
+}
+
+func (s *Apiserver) Destroy() error {
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err := os.Remove(recoveryPodManifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// TODO: consider removing the recoveryDir
+	// (The dir is tmp by default anyways and putting a wrong path here could be fatal as it would have to recursively delete.)
```

Comment:
which sensitive information (after recovery) does this dir contain?

Reply:
etcd credentials come to mind. Leaving those laying around is a no-go.

Reply:
they still have the same perms restriction as in etcd, I'll think about other options

Reply:
done